src/main/java/org/inferred/freebuilder/processor/PrimitiveOptionalProperty.java

```diff
@@ -0,0 +1,240 @@
+package org.inferred.freebuilder.processor;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.CaseFormat;
+import com.google.common.base.Optional;
+import org.inferred.freebuilder.processor.util.*;
+
+import javax.lang.model.type.TypeMirror;
+import java.util.OptionalDouble;
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+
+import static org.inferred.freebuilder.processor.BuilderMethods.*;
+import static org.inferred.freebuilder.processor.util.Block.methodBody;
+
+/**
+ * This property class handles the primitive optional fields, including
+ * {@link OptionalDouble}, {@link OptionalLong}, and {@link OptionalInt}.
+ */
+public class PrimitiveOptionalProperty extends PropertyCodeGenerator {
+    @VisibleForTesting
+    enum OptionalType {
+        INT(OptionalInt.class, "int"),
+        LONG(OptionalLong.class, "long"),
+        DOUBLE(OptionalDouble.class, "double");
+
+        private final QualifiedName cls;
+        private final String empty = "empty";
+        private final String ofNullable = "of";
+        private final String primitiveType;
+        private final String unaryType;
+        private final String unaryMethod;
+
+        OptionalType(Class<?> type, String primitiveType) {
+            this.cls = QualifiedName.of(type);
+            this.primitiveType = primitiveType;
+
+            String upperCamelName = CaseFormat.LOWER_CAMEL.to(CaseFormat.UPPER_CAMEL, primitiveType);
+            this.unaryType = upperCamelName + "UnaryOperator";
+            this.unaryMethod = "applyAs" + upperCamelName;
+        }
+
+        public static OptionalType lookup(TypeMirror elementType) {
+            String type = elementType.toString();
+
+            for (OptionalType op : values()) {
+                if (type.contains(op.cls.getSimpleName())) {
+                    return op;
+                }
+            }
+
+            throw new IllegalStateException("Not a supported type: " + type);
+        }
+
+        protected void applyMapper(SourceBuilder code, Metadata metadata, Metadata.Property property) {
+            code.addLine("  if(%s().isPresent()) {", getter(property));
```

Comment:
I think it would be more idiomatic to use ifPresent:

    property.ifPresent(value -> setProperty(mapper.apply(value)));